core/src/main/java/org/apache/calcite/sql/parser/SqlAbstractParserImpl.java

```diff
@@ -619,9 +619,11 @@ public static LexicalState forConfig(SqlParser.Config config) {
         return BQID;
       case BACK_TICK:
         if (config.conformance().allowHyphenInUnquotedTableName()
-            && config.charLiteralStyles().equals(
+            && (config.charLiteralStyles().equals(
                 EnumSet.of(CharLiteralStyle.BQ_SINGLE,
-                    CharLiteralStyle.BQ_DOUBLE))) {
+                    CharLiteralStyle.BQ_DOUBLE))
+               || config.charLiteralStyles().equals(
+            EnumSet.of(CharLiteralStyle.STANDARD)))) {
```

Comment:
With this change, any dialect whose `allowHyphenInUnquotedTableName` returns true will allow identifiers with back-ticks. I don't think anyone will expect that.

Reply:
then it is not clear what is the purpose of `allowHyphenInUnquotedTableName` returning `true` then?
Isn't this what is meant in javadoc
https://github.com/apache/calcite/blob/2fafa5254e40ddcffbf1fcd9ae56c50954763dac/core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java#L247-L250

Reply:
You should take the time to understand how it currently works. The implementation is subtle.

Lexical states are a very clever tool (I introduced them myself, to allow switching between quoting styles) but they are costly. Each lexical state creates a whole new copy of the parser's transition table and therefore makes the parser much larger.

Therefore we should only support the combinations that people actually need.

Relying on the sets of literal styles to deduce that we are dealing with BigQuery was a hack on my part. (See https://issues.apache.org/jira/browse/CALCITE-4247.) But that doesn't give you license to make the hack worse. In fact, if you want to change things, you have to figure out a way to make things better.